### ISSUE TITLE
Documentation for tensor5 manipulation

### ISSUE THREAD

**Issue by pfeatherstone** [Dobiasd/frugally-deep#149]
Support reshaping and concatenating fdeep::tensor5 tensors, i.e. manipulating tensors post inference

**Comment by pfeatherstone** [comment]
sorry, it looks like this is supported. There are a bunch of functions in tensor5.hpp. Maybe the issue should be changed to request documentation for tensor manipulation

**Comment by Dobiasd** [comment]
Thanks for the remark. At first, I thought you mean support for `reshape` and `concatenate` layers. But you mean doing this kind of stuff outside the model.
Yeah, right not all these things are in namespace `fdeep::internal`, because they are not part of the public API, but only used for the layer's implementations.

I'm not sure if it would be wise to expose all those things and thus increase the API surface that much, mainly because:
- It would give more room to break backwards compatibility and thus maybe pin down things to much.
- Up to now, frugally-deep was only intended to run `model.predict` of Keras models in C++, not to be a general tensor manipulation library. I'm not sure if it makes sense to go down this road.

What do you think?

**Comment by Dobiasd** [comment]
Oh, also you spiked my interest. Are you thinking of some special kind of Tensor manipulation that can not be done in `reshape`/`concatenate` layers in Keras? Because if not, I'd suggest to simply do these post-inference manipulations at the end of the model you export from Keras to fdeep. :)

**Comment by pfeatherstone** [comment]
I’m using the functions in fdeep::internal. It’s up to me to keep track of changes. I was trying out yolov3 inference from the Keras implementation out of curiosity. For yolov3 you have to do some reshaping, permuting and even activation post inference

**Comment by pfeatherstone** [comment]
Worth saying that if you were to use arrayfire as a backend, tensor manipulation would be very easy and you could simply expose its API rather than using fdeep::tensor5 objects

**Comment by Dobiasd** [comment]
> I’m using the functions in fdeep::internal. It’s up to me to keep track of changes. 

That's exactly the point. :slightly_smiling_face:
You can do it, but have to expect your code to break with future version of frugally-deep. Also such implementation details do not need to be documented to the public. I'd even argue that it's better to not document it, because documentation might make new users think that it's a stable and supported feature.

> I was trying out yolov3 inference from the Keras implementation out of curiosity. For yolov3 you have to do some reshaping, permuting and even activation post inference

Have you tried to implement the reshaping as additional Keras layers in Python and wrap your prediction model into them with the [functional API](https://keras.io/getting-started/functional-api-guide/)? If that is possible, you would have more compatibility guarantees.

---

> Worth saying that if you were to use arrayfire as a backend, tensor manipulation would be very easy and you could simply expose its API rather than using fdeep::tensor5 objects

That does not sound very good to me. Exposing arrayfire on the public API of frugally-deep would mean to be pinned down on it forever, or break much existing client code in case of future changes. And that is the exact reason why Eigen currently is not exposed. Just imagine I would have exposed it. In that case, we would not even have the freedom to actually replace it with arrayfire.

So, in conclusion, I think such an implementation detail (eigen/arrayfire) should not be carelessly made visible on the API surface of fdeep.

**Comment by Dobiasd** [comment]
Meta:

> So, in conclusion, I think such an implementation detail (eigen/arrayfire) should not be carelessly made visible on the API surface of fdeep

I guess we witness a cool instance of a quite general phenomenon here. :slightly_smiling_face:

From a user's perspective, "just exposing stuff" can be desired. It seems like a quick and easy win.

However, for the maintainer, who probably has to manage the project for a few more years, it can be more important to keep the complexity low and not impede possible future development.

**Comment by pfeatherstone** [comment]
>
 or break much existing client code in case of future changes
> 
I guess so. But, unless i'm mistaken, it's still your project and shouldn't have to care if it breaks other people's code. They should be thankful you've open sourced this in the first place and not complain when they code breaks even though they didn't have to do most of the leg work.

**Comment by Dobiasd** [comment]
Even though I am the main maintainer, I wouldn't say that it's just my project. Multiple people contributed. :)

Anyways, for me it's not about ownership or complains, but about providing a reliable tool that people can confidently build their production code against.

And one aspect of quality, for me, is avoiding breaking changes, at least when feasible, even though the library on purpose is released with an 0.x version number instead of 1.x.

Or, if you prefer the ownership approach, it's my library and I just don't want to expose Eugen/arrayfire. :-p

**Comment by pfeatherstone** [comment]
Fair enough. That's the spirit. Anyways, it's been very useful for me. FYI, yolov3 inference using https://github.com/qqwweee/keras-yolo3 and your library takes about 1.1s on a single image of size 416x416x3 with -O3 -ffast-math and -mavx2 optimizations. It's possible i'm not doing tensor manipulation in the most efficient away but that's still better than the darknet repository...

**Comment by Dobiasd** [comment]
> Fair enough. That's the spirit.

Hehe, guessed you'd like it. ;)

---

Anyways: Have you tried to implement the reshaping as additional Keras layers in Python and wrap your prediction model into them with the functional API?

I mean something like that:

```python3
from keras.layers import Input, Conv2D, Reshape
from keras.models import Model


def main_model():
    inputs = Input(shape=(416, 416, 3))
    x = Conv2D(4, (3, 3), padding='same')(inputs)
    # Whatever complex stuff is here
    x = Conv2D(8, (3, 3), padding='same')(x)
    model = Model(inputs=inputs, outputs=x)
    model.compile(loss='categorical_crossentropy', optimizer='nadam')
    return model


inner_model = main_model()

# Wrap the inner model into a second model,
# adding the post-prediction reshape operations as layers.
outer_inputs = Input(shape=(256, 256, 3))
x = inner_model(outer_inputs)
x = Reshape((128, 128, 32))(x)
wrapping_model = Model(inputs=outer_inputs, outputs=x)

wrapping_model.save('wrapping_model.h5', include_optimizer=False)
```

**Comment by pfeatherstone** [comment]
I could do but it would be a bit of a distraction. I probably would use Opencv for yolo inference since it’s very fast and I would be doing computer vision stuff anyway. I was mainly curious as to how fdeep worked. I’m certainly going to use it in the future though. I love header only libraries.

**Comment by Dobiasd** [comment]
Thanks for the interesting discussion, and good luck with your project. Please let me know about further issues. :)
